### DISCUSSION TITLE
Handling on events

### DISCUSSION THREAD

**Discussion by cdlus** [a-h/templ#228] [Q&A]
How do you pass the event to the script functions when doing `onClick{Something()}` ?

**Comment by a-h** [comment]
The current strategy would be to use standard JavaScript instead of script templates.

You can use the "constant" version of HTML attributes to write any JavaScript you want.

```html
<input id="btn" type="button" onclick="(e) => alert(e)" value="Click" />
```

Or you can use a `<script>` element to access the element's `addEventListener` function, e.g. 

```html
<script>
document.getElementById("btn").addEventListener("click", greet);
</script>
```

However, I think that a small extension to the script template processing could satisfy the use case for script templates.

We'd add a new `JSVar` type. This type can be used as an input for a script template that isn't known at compile time - i.e. we expect it to be in scope within the JavaScript.

Then, a convenience variable of `Event`.

```go
// JSVar is a JavaScript variable. templ will not escape this value.
// The value must be a valid JavaScript identifier that is available at
// runtime.
type JSVar struct {
	Name string
}

// Event is a JavaScript event variable, commonly used in event handlers.
var Event JSVar = JSVar{Name: "event"}

// SafeScript encodes unknown parameters for safety.
func SafeScript(functionName string, params ...any) string {
	encodedParams := make([]string, len(params))
	for i := 0; i < len(encodedParams); i++ {
		if v, isJSVar := params[i].(JSVar); isJSVar {
			encodedParams[i] = v.Name
			continue
		}
		enc, _ := json.Marshal(params[i])
		encodedParams[i] = EscapeString(string(enc))
	}
	sb := new(strings.Builder)
	sb.WriteString(functionName)
	sb.WriteRune('(')
	sb.WriteString(strings.Join(encodedParams, ","))
	sb.WriteRune(')')
	return sb.String()
}
```

In the template, you could then do this:

```html
script alert(e templ.JSVar) {
  alert("Event: " + e);
  alert("Target: " + e.target);
}

templ button(global, user int) {
	<input type="button" value="Do something" onclick={ alert(templ.Event) }/>
}
```

The output is rendered as:

```html
<script type="text/javascript">
function __templ_alert_3592(e){
  alert("Event: " + e);
  alert("Target: " + e.target);
}
</script>
<input type="button" value="Do something" onclick="__templ_alert_3592(event)"></div>
```

I'd love to hear your thoughts on that.

**Comment by a-h** [comment]
Just following up on this.

Given the following templ template:

```templ
package components

script onClick(name string) {
	alert("Hello " + name + " from " + event.type + " event");
}

templ Button(name string) {
	<button onclick={ onClick(name) }>Click me ({ name })</button>
}

templ Page() {
	<!DOCTYPE html>
	<html>
		<head>
			<title>Buttons</title>
		</head>
		<body>
			<div>
				@Button("Adam")
			</div>
			<div>
				@Button("Bob")
			</div>
			<div>
				@Button("Charlie")
			</div>
		</body>
	</html>
}
```

And a `main.go`:

```go
package main

import (
	_ "embed"
	"net/http"
	"onclick/components"

	"github.com/a-h/templ"
)

func main() {
	component := components.Page()

	handler := templ.Handler(component)

	http.Handle("/", handler)
	http.ListenAndServe("localhost:8080", nil)
}
```

You can access the `event` without any gymnastics:

<img width="1017" alt="Screenshot 2024-01-05 at 21 30 30" src="https://github.com/a-h/templ/assets/1029947/1fb59c5a-b83a-44a0-a366-ab3b3fe27590">

The HTML that's output is as follows:

```html
<!doctype html>
<html>
    <head>
        <title>Buttons</title>
    </head>
    <body>
        <div>
            <script type="text/javascript">
                function __templ_onClick_f1dd(name) {
                    alert("Hello " + name + " from " + event.type + " event");
                }
            </script>
            <button onclick="__templ_onClick_f1dd(&#34;Adam&#34;)">Click me (Adam)</button>
        </div>
        <div>
            <button onclick="__templ_onClick_f1dd(&#34;Bob&#34;)">Click me (Bob)</button>
        </div>
        <div>
            <button onclick="__templ_onClick_f1dd(&#34;Charlie&#34;)">Click me (Charlie)</button>
        </div>
    </body>
</html>
``` 

It works in Safari too.

<img width="1129" alt="Screenshot 2024-01-05 at 21 35 42" src="https://github.com/a-h/templ/assets/1029947/091d9049-d700-4486-86a2-cdf31a5ddca2">

If you're uncomfortable with the fact that event is magically available in the function, you can rewrite the component script output to add the event parameter, or (better) make a ComponentScript of your own. I think it would be better if the Component Script output kept the FunctionName and Arguments separate, since it would make the change not so awkward as using a find/replace less brittle, so I might change up the `templ.ComponentScript` in the future.

Here's the example code for that.

```templ
func EventHandler(cs templ.ComponentScript) templ.ComponentScript {
	newName := cs.Name + "_EventHandler"
	newFunction := strings.Replace(cs.Function, cs.Name+"(", newName+"(event, ", 1)
	newCall := strings.Replace(cs.Call, cs.Name+"(", newName+"(event, ", 1)
	return templ.ComponentScript{
		Name:       newName,
		Function:   newFunction,
		Call:       newCall,
		CallInline: newCall,
	}
}

script onClick(name string) {
	alert("Hello " + name + " from " + event.type + " event");
}

templ Button(name string) {
	<button onclick={ EventHandler(onClick(name)) }>Click me ({ name })</button>
}
```

I'll close this discussion... Feel free to re-open if there's more to add!

**Reply by claritydistro-nick** [reply]
I would suggest adding this to the documentation if at all possible. How to actually handle events when using script templates is pretty hard to figure out given the magic event piece.

**Reply by a-h** [reply]
Great idea. I'd take a PR for that if you have time?

The docs are markdown in the repo, see the /docs directory.
